compiler/rustc_typeck/src/astconv/generics.rs

```diff
@@ -336,204 +339,148 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
     pub(crate) fn check_generic_arg_count(
         tcx: TyCtxt<'_>,
         span: Span,
-        def: &ty::Generics,
-        args: &hir::GenericArgs<'_>,
-        position: GenericArgPosition,
+        def_id: DefId,
+        gen_params: &ty::Generics,
+        gen_args: &hir::GenericArgs<'_>,
+        gen_pos: GenericArgPosition,
         has_self: bool,
         infer_args: bool,
     ) -> GenericArgCountResult {
```

Comment:
I'm looking around for ways to reduce the amount of span wrangling that we need to do here, and it seems to me that we could make `check_generic_arg_count` receive a `PathSegment`, that would simplify the work needed to be done to find the `::` and `<` for accurate spans. The only call to this method that doesn't have direct access to the correct `PathSegment` is in `create_substs_for_ast_path`, but _that_ method is called from a place that has access to one, except for lang items, for which we can either build a face `PathSegment` or make the arg an `Option` or something.

Reply:
done 👍🏻